File: python/Tries/design_add_search_words_data_structure.py
class TrieNode:
    """
    Represents a node in the Trie data structure for Word Dictionary.
    
    Attributes:
        children (dict): Dictionary mapping characters to child TrieNodes
        word (bool): Flag indicating if this node represents the end of a word
    """
    def __init__(self):
        self.children = {} 
        self.word = False

class WordDictionary:
    """
    Implementation of a Word Dictionary that supports adding words and searching with wildcards.
    
    This data structure extends the basic Trie to support pattern matching with '.' wildcards.
    It's useful for:
    - Autocomplete with partial matching
    - Pattern matching in text processing
    - Word games like Wordle
    - Search functionality with wildcards
    
    Time Complexity:
        - addWord: O(m) where m is the length of the word
        - search: O(26^m) in worst case where m is the length of the word
                 (due to wildcard '.' that can match any character)
                 O(m) in best case when no wildcards are present
    
    Space Complexity:
        - O(ALPHABET_SIZE * N * M) where:
          - ALPHABET_SIZE is the number of possible characters (26 for lowercase letters)
          - N is the number of words
          - M is the average length of words
        - In worst case: O(ALPHABET_SIZE * N * M) when no words share prefixes
        - In best case: O(ALPHABET_SIZE * M) when all words share the same prefix
    """

    def __init__(self):
        """
        Initialize an empty Word Dictionary with a root node.
        
        Time Complexity: O(1)
        Space Complexity: O(1)
        """
        self.root = TrieNode()

    def addWord(self, word: str) -> None:
        """
        Add a word to the dictionary.
        
        Algorithm:
        1. Start from the root node
        2. For each character in the word:
           - If the character doesn't exist as a child, create a new TrieNode
           - Move to the child node
        3. Mark the final node as the end of a word
        
        Args:
            word (str): The word to add to the dictionary
            
        Time Complexity: O(m) where m is the length of the word
        Space Complexity: O(m) in worst case when no existing path exists
        """
        curr = self.root

        for c in word:
            if c not in curr.children:
                curr.children[c] = TrieNode()
            curr = curr.children[c]
        curr.word = True

    def search(self, word: str) -> bool:
        """
        Search for a word in the dictionary. Supports '.' wildcard that matches any character.
        
        Algorithm:
        1. Use DFS to traverse the Trie
        2. For each character in the search pattern:
           - If it's a '.', try all possible children recursively
           - If it's a regular character, follow that specific path
        3. Return True if we reach the end and the final node is marked as end of word
        
        Args:
            word (str): The word/pattern to search for (can contain '.' wildcards)
            
        Returns:
            bool: True if the word/pattern exists in the dictionary, False otherwise
            
        Time Complexity: O(26^m) in worst case where m is the length of the word
                        (when the word contains many '.' wildcards)
                        O(m) in best case when no wildcards are present
        Space Complexity: O(m) due to recursion stack depth
        """
        def dfs(j, root):
            curr = root
            
            for i in range(j, len(word)):
                c = word[i]

                if c == ".":
                    # Wildcard case: try all possible children
                    for child in curr.children.values():
                        if dfs(i + 1, child):
                            return True
                    return False
                else:
                    # Regular character case: follow specific path
                    if c not in curr.children:
                        return False
                    curr = curr.children[c]
            return curr.word
        
        return dfs(0, self.root)

# -------------------
# Test Cases
# -------------------
def test_word_dictionary():
    wd = WordDictionary()
    # Add words
    wd.addWord("bad")
    wd.addWord("dad")
    wd.addWord("mad")
    wd.addWord("pad")
    wd.addWord("baddy")
    wd.addWord("")  # Add empty string
    # Exact matches
    assert wd.search("bad") == True
    assert wd.search("dad") == True
    assert wd.search("mad") == True
    assert wd.search("pad") == True
    assert wd.search("baddy") == True
    assert wd.search("") == True
    # Not present
    assert wd.search("badd") == False
    assert wd.search("bads") == False
    assert wd.search("b") == False
    # Wildcard matches
    assert wd.search(".ad") == True
    assert wd.search("b..") == True
    assert wd.search("..d") == True
    assert wd.search("baddy") == True
    assert wd.search("ba..y") == True
    assert wd.search(".....") == True
    assert wd.search("..") == False
    # Wildcard not present
    assert wd.search("..z") == False
    # Edge cases
    assert wd.search(".") == False
    assert wd.search("....") == False
    print("All test cases passed!")

if __name__ == "__main__":
    test_word_dictionary()

File: C++/Tries/design_add_search_words_data_structure.cpp
#include <bits/stdc++.h>
using namespace std;

// TrieNode class to represent each node of the Trie
class TrieNode {
public:
    unordered_map<char, TrieNode*> children; // maps character to next node
    bool word; // marks if this node represents end of a word

    TrieNode() {
        word = false; // initially not the end of any word
    }
};

class WordDictionary {
private:
    TrieNode* root;

    /**
     * Recursive DFS helper to search for a word/pattern.
     * Handles '.' wildcard which matches any character.
     */
    bool dfs(int j, string& word, TrieNode* curr) {
        for (int i = j; i < word.size(); ++i) {
            char c = word[i];

            if (c == '.') {
                // Wildcard: try all possible children
                for (auto& p : curr->children) {
                    TrieNode* child = p.second;
                    if (dfs(i + 1, word, child)) {
                        return true;
                }   
}
                return false; // no matching path
            } else {
                // Regular character
                if (curr->children.find(c) == curr->children.end()) {
                    return false; // path does not exist
                }
                curr = curr->children[c];
            }
        }
        return curr->word; // reached end of word?
    }

public:
    /**
     * WordDictionary: Trie with support for '.' wildcard in search.
     * 
     * Time Complexity:
     *  - addWord: O(m), m = length of the word
     *  - search: O(26^m) worst case (due to wildcards), O(m) best case
     * 
     * Space Complexity:
     *  - Worst: O(ALPHABET_SIZE * N * M)
     */

    WordDictionary() {
        root = new TrieNode(); // initialize root node
    }

    /**
     * Adds a word into the Trie.
     * Time: O(m)
     * Space: O(m) in worst case when creating all new nodes
     */
    void addWord(string word) {
        TrieNode* curr = root;

        for (char c : word) {
            if (curr->children.find(c) == curr->children.end()) {
                curr->children[c] = new TrieNode();
            }
            curr = curr->children[c];
        }
        curr->word = true; // mark end of word
    }

    /**
     * Searches for a word or pattern (with '.' wildcard).
     * Time: O(26^m) worst case, O(m) best case
     * Space: O(m) due to recursion
     */
    bool search(string word) {
        return dfs(0, word, root);
    }
};


/*
=======================================================================
Line-by-line Explanation:
-----------------------------------------------------------------------

#include <bits/stdc++.h>
- Includes all standard C++ libraries.

using namespace std;
- Avoids writing std:: repeatedly.

class TrieNode {
- Defines a TrieNode that holds children and a flag for end-of-word.

unordered_map<char, TrieNode*> children;
- Maps a character to its child TrieNode.

bool word;
- Indicates if this node is the end of a valid word.

TrieNode() { word = false; }
- Constructor initializes word = false.

class WordDictionary {
- Defines the WordDictionary class with all required methods.

TrieNode* root;
- Pointer to the root of the Trie.

bool dfs(int j, string& word, TrieNode* curr)
- Helper function to recursively search for the word.
- Handles '.' by iterating through all children.

WordDictionary() {
  root = new TrieNode();
}
- Initializes the root node when creating a WordDictionary.

void addWord(string word)
- Inserts a word into the Trie.

bool search(string word)
- Searches for a word in the Trie, calling dfs().

for (char c : word)
- Range-based loop over characters of the word.

if (curr->children.find(c) == curr->children.end())
- Checks if character c exists in children map.

curr->children[c] = new TrieNode();
- Creates a new TrieNode if not present.

return dfs(0, word, root);
- Starts DFS traversal at index 0 from root.

for (auto& p : curr->children) 
    // Iterates over all children nodes (p.second is the child TrieNode*).

=======================================================================
Time Complexity:
-----------------------------------------------------------------------
addWord:  O(m), where m = length of word.
search:   O(26^m) worst case (due to wildcards), O(m) best case.

Space Complexity:
-----------------------------------------------------------------------
- Worst case: O(ALPHABET_SIZE * N * M)
  where:
  ALPHABET_SIZE = number of possible characters (26 for lowercase),
  N = number of words,
  M = average length of words.
- Recursion stack in search adds O(m) space.

=======================================================================
*/

